src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/MyServices/Internal/ContextValue.vb

```diff
@@ -0,0 +1,115 @@
+
+            If HttpContextType IsNot Nothing Then
+                Return HttpContextType.GetProperty("Current")
+            Else
+                Return Nothing
+            End If
+        End Function
+
+        'This class isn't meant to be constructed.
+        Private Sub New()
+        End Sub
+
+        Private Shared m_HttpContextCurrent As System.Reflection.PropertyInfo = InitContext()
+    End Class
+
+    '''**************************************************************************
+    ''' ;ContextValue
+    ''' <summary>
+    ''' Stores an object in a context appropriate for the environment we are 
+    ''' running in (web/windows)
+    ''' </summary>
+    ''' <typeparam name="T"></typeparam>
+    ''' <remarks>
+    ''' "Thread appropriate" means that if we are running on ASP.Net the object will be stored in the 
+    ''' context of the current request (meaning the object is stored per request on the web).  Otherwise, 
+    ''' the object is stored per CallContext.  Note that an instance of this class can only be associated
+    ''' with the one item to be stored/retrieved at a time.
+    ''' </remarks>
+    <Global.System.ComponentModel.EditorBrowsableAttribute(Global.System.ComponentModel.EditorBrowsableState.Never)>
+    Public Class ContextValue(Of T)
+        Public Sub New()
+            m_ContextKey = System.Guid.NewGuid.ToString
+        End Sub
+
+        '''**************************************************************************
+        ''' ;Value
+        ''' <summary>
+        ''' Get the object from the correct thread-appropriate location
+        ''' </summary>
+        ''' <value></value>
+        Public Property Value() As T 'No Synclocks required because we are operating upon instance data and the object is not shared across threads
+            Get
+                Dim items As IDictionary = GetDictionary()
+                Return DirectCast(items(m_ContextKey), T) 'Note, IDictionary(key) can return Nothing and that's ok
+            End Get
+            Set(ByVal value As T)
+                Dim items As IDictionary = GetDictionary()
+                items(m_ContextKey) = value
+            End Set
+        End Property
+
+        Private Shared Function GetDictionary() As IDictionary
+            Dim context = SkuSafeHttpContext.Current()
+            If context IsNot Nothing Then
+                Return context.Items
+            End If
+            If s_ThreadLocal Is Nothing Then
+                Interlocked.CompareExchange(s_ThreadLocal, New ThreadLocal(Of IDictionary)(Function() New Dictionary(Of String, T)), Nothing)
+            End If
+            Return s_ThreadLocal.Value
+        End Function
+
+        Private ReadOnly m_ContextKey As String 'An item is stored in the dictionary by a guid which this string maintains
+
+        Private Shared s_ThreadLocal As ThreadLocal(Of IDictionary)
```

Comment:
Would AsyncLocal be better?

Reply:
I'm not sure if I understand all the details, but it seems like you could use AsyncLocal always and drop the HttpContext stuff.

Reply:
I've not read the rest of the PR yet, but if the goal is to replace CallContext.LogicalGet/SetData, AsyncLocal is a much better fit than ThreadLocal.

Reply:
Ah, it appears the goal is to replace CallContext.GetData/SetData, rather than LogicalGetData/SetData, so ThreadLocal is probably right.